File: src/two_sum_1/solution_binary_search.go
package two_sum_1

import (
	"fmt"
	"sort"
)

type Pair struct {
	num int
	idx int
}

func twoSumBinarySearch(nums []int, target int) []int {
	numsWithIndices := make([]Pair, len(nums))
	for i, num := range nums {
		numsWithIndices[i] = Pair{num: num, idx: i}
	}

	sort.Slice(numsWithIndices, func(i, j int) bool {
		return numsWithIndices[i].num < numsWithIndices[j].num
	})

	for idx, pair := range numsWithIndices {
		num := pair.num
		originalIdx := pair.idx
		find := target - num

		l := idx + 1
		r := len(numsWithIndices)
		for l < r {
			m := (l + r) / 2
			if numsWithIndices[m].num == find {
				return []int{originalIdx, numsWithIndices[m].idx}
			} else if numsWithIndices[m].num < find {
				l = m + 1
			} else {
				r = m
			}
		}
	}

	return []int{-1, 1}
}

func TestBinarySearch() {
	fmt.Println(twoSumBinarySearch([]int{3, 2, 4}, 6))
}

File: src/two_sum_1/solution_binary_search.ts
interface Pair {
  num: number;
  idx: number;
}

function twoSum(nums: number[], target: number): number[] {
  const numsWithIndices: Pair[] = nums
    .map((num, idx) => ({ num, idx }))
    .sort((a, b) => a.num - b.num);

  for (let idx = 0; idx < numsWithIndices.length; idx++) {
    const { num, idx: originalIdx } = numsWithIndices[idx];
    const find = target - num;

    let l = idx + 1;
    let r = numsWithIndices.length;

    while (l < r) {
      const m = Math.floor((l + r) / 2);

      if (numsWithIndices[m].num === find) {
        return [originalIdx, numsWithIndices[m].idx];
      } else if (numsWithIndices[m].num < find) {
        l = m + 1;
      } else {
        r = m;
      }
    }
  }

  return [-1, -1];
}

export function test() {
  console.log(twoSum([3, 2, 4], 6));
}
